python/cuml/manifold/t_sne.pyx

```diff
@@ -0,0 +1,472 @@
+        self.post_momentum = post_momentum
+        self.learning_rate_method = learning_rate_method
+
+        self.epssq = 0.0025
+        self.perplexity_tol = 1e-5
+        self.min_gain = 0.01
+        self.pre_learning_rate = learning_rate
+        self.post_learning_rate = learning_rate * 2
+
+        self._should_downcast = should_downcast
+        self.Y = None
+        return
+
+    def __repr__(self):
+        """
+        Pretty prints the arguments of a class using Sklearn standard :)
+        """
+        cdef list signature = inspect.getfullargspec(self.__init__).args
+        if signature[0] == 'self':
+            del signature[0]
+        cdef dict state = self.__dict__
+        cdef str string = self.__class__.__name__ + '('
+        cdef str key
+        for key in signature:
+            if key not in state:
+                continue
+            if type(state[key]) is str:
+                string += "{}='{}', ".format(key, state[key])
+            else:
+                if hasattr(state[key], "__str__"):
+                    string += "{}={}, ".format(key, state[key])
+        string = string.rstrip(', ')
+        return string + ')'
+
+    def fit(self, X):
+        """Fit X into an embedded space.
+        Parameters
+        ----------
+        X : array-like (device or host) shape = (n_samples, n_features)
+                X contains a sample per row.
+                Acceptable formats: cuDF DataFrame, NumPy ndarray, Numba device
+                ndarray, cuda array interface compliant array like CuPy
+        y : array-like (device or host) shape = (n_samples, 1)
+                y contains a label per row.
+                Acceptable formats: cuDF Series, NumPy ndarray, Numba device
+                ndarray, cuda array interface compliant array like CuPy
+        """
+        cdef int n, p
+        cdef cumlHandle* handle_ = <cumlHandle*><size_t>self.handle.getHandle()
+        assert(handle_ != NULL)
+
+        if len(X.shape) != 2:
+            raise ValueError("data should be two dimensional")
+
+        cdef uintptr_t X_ptr
+        if self._should_downcast:
+            _X, X_ptr, n, p, dtype = to_cuda(X, order='C',
+                                             convert_to_dtype=np.float32)
+        else:
+            _X, X_ptr, n, p, dtype = to_cuda(X, order='C',
+                                             check_dtype=np.float32)
+
+        if n <= 1:
+            raise ValueError("There needs to be more than 1 sample to build "
+                             "nearest the neighbors graph")
+
+        self.n_neighbors = min(n, self.n_neighbors)
+        if self.perplexity > n:
+            print("[Warn] Perplexity = {} should be less than the "
+                  "# of datapoints = {}.".format(self.perplexity, n))
+            self.perplexity = n
+
+        # Prepare output embeddings
+        Y = cuda.device_array(
+            (n, self.n_components),
+            order="F",
```

Comment:
Curiosity question, why is `Y` `F` major while `X` is `C` major?

Reply:
Good Q! X is C major because I think KNN only accepts as of now C order (might change later). TSNE is much faster on F contiguous arrays :) Hence F Contiguous. (TSNE's data access patterns are column wise)